File: apps/server/api/src/services/fs.service.ts
/**
 * Filesystem & File Browser Service.
 * Implements business logic for browsing system directories, reading/writing files, creating/deleting folders.
 */
import fs from "node:fs/promises";
import os from "node:os";
import path from "node:path";
import { listDirTool, readFileTool, writeFileTool } from "@/agent/src/tools/index.js";
import {
  IMAGE_MAX_BYTES,
  MAX_FILE_PREVIEW_BYTES,
  formatBytes,
  getFilePreviewBlock,
  imageMimeForExtension,
  isLockFileName,
  isPreviewableImageName,
  isSvgFileName,
} from "@console/types";
import type { FilePreviewBlockedCode } from "@console/types";
import type { FsTreeEntry } from "@console/types";
import { isPathIgnored } from "@/api/src/utils/ignored.js";
import { isHiddenName, safeFileSize, toSortedEntries } from "@/agent/src/tools/fs-common.js";

/**
 * Raised when a file may not be returned as a text preview (lockfile, binary,
 * or over the size cap). Routes translate this into a structured HTTP error.
 */
export class FilePreviewBlockedError extends Error {
  readonly code: FilePreviewBlockedCode;
  readonly status: 413 | 415;
  /** Extra structured fields for the JSON error body. */
  readonly detail?: Record<string, number>;

  constructor(
    code: FilePreviewBlockedCode,
    message: string,
    options?: { status?: 413 | 415; detail?: Record<string, number> },
  ) {
    super(message);
    this.name = "FilePreviewBlockedError";
    this.code = code;
    this.status = options?.status ?? 415;
    this.detail = options?.detail;
  }
}

/** Bytes inspected for NUL bytes before trusting that a file is text. */
const BINARY_SNIFF_BYTES = 8192;

/** Weak ETag from size + mtime so repeated previews get 304s. */
export function buildFileETag(sizeBytes: number, mtimeMs: number): string {
  return `W/"${sizeBytes.toString(36)}-${Math.round(mtimeMs).toString(36)}"`;
}

/**
 * Slice already-decoded text to [startLine, endLine] (1-based, inclusive)
 * with a single scan — avoids splitting the whole file into an array.
 * Matches the previous split/slice/join semantics exactly (including
 * trailing-newline handling).
 */
function sliceLines(text: string, startLine?: number, endLine?: number): string {
  if (startLine === undefined && endLine === undefined) return text;
  // Line count as split("\n").length without materialising the array.
  let count = 1;
  for (let i = 0; i < text.length; i++) if (text.charCodeAt(i) === 10) count++;
  const start0 = Math.max(1, startLine ?? 1) - 1;
  const endExcl = Math.min(endLine ?? count, count);
  if (start0 >= count || endExcl <= start0) return "";
  // Locate the offset where line `idx` (0-based) starts.
  const offsetOf = (idx: number): number => {
    if (idx <= 0) return 0;
    let line = 0;
    for (let i = 0; i < text.length; i++) {
      if (text.charCodeAt(i) === 10) {
        line++;
        if (line === idx) return i + 1;
      }
    }
    return text.length;
  };
  const sliceStart = offsetOf(start0);
  if (endExcl >= count) return text.slice(sliceStart);
  return text.slice(sliceStart, offsetOf(endExcl) - 1);
}

export class FsService {
  /**
   * Browse a system directory for the mobile/desktop file picker UI.
   * Defaults to user's home directory ($HOME) when no path is provided.
   */
  async browseDirectory(
    targetPath?: string,
    showHidden = false,
  ): Promise<{ currentPath: string; parentPath: string | null; entries: FsTreeEntry[] }> {
    const home = os.homedir();
    const resolvedPath = targetPath ? path.resolve(targetPath) : home;
    const parentPath =
      path.dirname(resolvedPath) !== resolvedPath ? path.dirname(resolvedPath) : null;

    const dirEntries = await fs.readdir(resolvedPath, { withFileTypes: true });
    const entries = await toSortedEntries(resolvedPath, dirEntries, { showHidden });

    return {
      currentPath: resolvedPath,
      parentPath,
      entries,
    };
  }

  /**
   * Get formatted directory tree for a project path.
   */
  async getDirectoryTree(targetPath: string, maxDepth = 3, showHidden = false): Promise<string> {
    const result = (await listDirTool.execute(
      listDirTool.inputSchema.parse({
        path: targetPath,
        maxDepth,
        recursive: true,
        showHidden,
      }),
    )) as { content: Array<{ text: string }> };

    return result.content[0]?.text ?? "";
  }

  /**
   * List all file and directory entries under a project path recursively.
   * Returns a flat list (like T3's ProjectEntry[]) for building a client-side
   * file tree. Used by the mobile file browser for search and full-tree views.
   *
   * Parallelised: stats for a directory resolve together and subdirectories
   * are walked concurrently (bounded), instead of one stat + one awaited
   * recursion at a time. `withSizes: false` skips all stats for the fastest
   * tree-only scan; `maxEntries` caps runaway trees.
   */
  async listAllEntries(
    targetPath: string,
    maxDepth = 25,
    showHidden = false,
    options?: { withSizes?: boolean; maxEntries?: number },
  ): Promise<FsTreeEntry[]> {
    const resolvedRoot = path.resolve(targetPath);
    const withSizes = options?.withSizes ?? true;
    const maxEntries = options?.maxEntries ?? 30000;
    const counter = { count: 0 };
    const truncated = { value: false };

    const walk = async (currentPath: string, depth: number): Promise<FsTreeEntry[]> => {
      if (depth > maxDepth || truncated.value) return [];
      let dirEntries: import("node:fs").Dirent[];
      try {
        dirEntries = await fs.readdir(currentPath, { withFileTypes: true });
      } catch {
        return [];
      }
      const visible = dirEntries.filter((e) => showHidden || !isHiddenName(e.name));
      // Directories first, then files, alphabetically (numeric-aware).
      visible.sort((a, b) => {
        if (a.isDirectory() !== b.isDirectory()) return a.isDirectory() ? -1 : 1;
        return a.name.localeCompare(b.name, undefined, { numeric: true });
      });

      // Stats for the whole directory resolve together, not one by one.
      const sizes = withSizes
        ? await Promise.all(
            visible.map((entry) => {
              if (!showHidden && isPathIgnored(entry.name)) return Promise.resolve(undefined);
              const entryPath = path.join(currentPath, entry.name);
              return safeFileSize(entryPath, entry.isDirectory());
            }),
          )
        : visible.map(() => undefined);

      // Collect subdirectories to walk concurrently (bounded), then
      // interleave children DFS-style to preserve the previous ordering:
      // each directory immediately followed by its subtree.
      const subdirPaths: string[] = [];
      for (const entry of visible) {
        if (!showHidden && isPathIgnored(entry.name)) continue;
        if (entry.isDirectory()) subdirPaths.push(path.join(currentPath, entry.name));
      }
      const childrenByPath = new Map<string, FsTreeEntry[]>();
      const CONCURRENCY = 8;
      for (let i = 0; i < subdirPaths.length; i += CONCURRENCY) {
        if (truncated.value) break;
        const batch = subdirPaths.slice(i, i + CONCURRENCY);
        const children = await Promise.all(batch.map((p) => walk(p, depth + 1)));
        batch.forEach((p, j) => childrenByPath.set(p, children[j]!));
      }

      const local: FsTreeEntry[] = [];
      for (let i = 0; i < visible.length; i++) {
        if (truncated.value || counter.count >= maxEntries) {
          truncated.value = true;
          break;
        }
        const entry = visible[i]!;
        if (!showHidden && isPathIgnored(entry.name)) {
          const dirPath = path.join(currentPath, entry.name);
          local.push({ name: entry.name, path: dirPath, isDir: true });
          counter.count++;
          continue;
        }

        const entryPath = path.join(currentPath, entry.name);
        const isDir = entry.isDirectory();
        local.push({ name: entry.name, path: entryPath, isDir, size: sizes[i] });
        counter.count++;
        if (isDir) {
          // Children already counted inside the recursive walk.
          local.push(...(childrenByPath.get(entryPath) ?? []));
        }
      }
      return local;
    };

    return walk(resolvedRoot, 1);
  }

  /**
   * Read file content with line range support.
   *
   * Enforces the file-preview gate (docs/notes/file-preview-gating.md):
   * lockfiles, binary files (extension or NUL-byte sniff), and files over
   * MAX_FILE_PREVIEW_BYTES are rejected with FilePreviewBlockedError before
   * the body is read. Single open/read: stat + sniff + body share one
   * file handle instead of stat + open/sniff/close + reopen.
   */
  async readFileContentWithMeta(
    filePath: string,
    startLine?: number,
    endLine?: number,
  ): Promise<{ content: string; sizeBytes: number; mtimeMs: number }> {
    const fileName = path.basename(filePath);

    const nameBlock = getFilePreviewBlock(fileName);
    if (nameBlock?.kind === "LOCKFILE_BLOCKED") {
      throw new FilePreviewBlockedError("LOCKFILE_BLOCKED", nameBlock.message);
    }
    if (nameBlock?.kind === "BINARY_FILE") {
      throw new FilePreviewBlockedError("BINARY_FILE", nameBlock.message);
    }

    const handle = await fs.open(filePath, "r");
    try {
      const stat = await handle.stat();
      if (!stat.isFile()) {
        throw new Error(`${filePath} is not a regular file.`);
      }
      if (stat.size > MAX_FILE_PREVIEW_BYTES) {
        throw new FilePreviewBlockedError(
          "FILE_TOO_LARGE",
          `"${fileName}" is ${formatBytes(stat.size)} — previews are capped at ${formatBytes(MAX_FILE_PREVIEW_BYTES)}.`,
          { status: 413, detail: { sizeBytes: stat.size, maxBytes: MAX_FILE_PREVIEW_BYTES } },
        );
      }

      // Single read: sniff the head of the same buffer for NUL bytes.
      // Extensions lie, NUL bytes don't — catches binaries whose extension
      // isn't on the denylist.
      const buffer = Buffer.alloc(stat.size);
      await handle.read(buffer, 0, stat.size, 0);
      const headLen = Math.min(BINARY_SNIFF_BYTES, buffer.length);
      for (let i = 0; i < headLen; i++) {
        if (buffer[i] === 0) {
          throw new FilePreviewBlockedError(
            "BINARY_FILE",
            `"${fileName}" doesn't look like a text file.`,
          );
        }
      }

      const raw = buffer.toString("utf-8");
      return { content: sliceLines(raw, startLine, endLine), sizeBytes: stat.size, mtimeMs: stat.mtimeMs };
    } finally {
      await handle.close();
    }
  }

  async readFileContent(filePath: string, startLine?: number, endLine?: number): Promise<string> {
    return (await this.readFileContentWithMeta(filePath, startLine, endLine)).content;
  }

  /**
   * Validate an image/SVG preview request and return its metadata.
   * Lets the route stream `Bun.file()` (zero-copy sendfile) instead of
   * buffering the whole file through Node.
   */
  async getImageMeta(
    filePath: string,
  ): Promise<{ mimeType: string; sizeBytes: number; mtimeMs: number }> {
    const fileName = path.basename(filePath);

    if (isLockFileName(fileName)) {
      throw new FilePreviewBlockedError(
        "LOCKFILE_BLOCKED",
        `"${fileName}" is a generated lockfile — open it on your machine instead.`,
      );
    }

    if (!isPreviewableImageName(fileName) && !isSvgFileName(fileName)) {
      throw new FilePreviewBlockedError(
        "BINARY_FILE",
        `"${fileName}" is not a supported image format.`,
      );
    }

    const stat = await fs.stat(filePath);
    if (!stat.isFile()) {
      throw new Error(`${filePath} is not a regular file.`);
    }

    if (stat.size > IMAGE_MAX_BYTES) {
      throw new FilePreviewBlockedError(
        "FILE_TOO_LARGE",
        `"${fileName}" is ${formatBytes(stat.size)} — image previews are capped at ${formatBytes(IMAGE_MAX_BYTES)}.`,
        { status: 413, detail: { sizeBytes: stat.size, maxBytes: IMAGE_MAX_BYTES } },
      );
    }

    const mimeType = imageMimeForExtension(path.extname(filePath)) ?? "application/octet-stream";
    return { mimeType, sizeBytes: stat.size, mtimeMs: stat.mtimeMs };
  }

  /**
   * Read raw image / SVG bytes for binary preview.
   *
   * Only allows supported raster image extensions and SVG.
   * Lockfiles and non-image binaries/text are blocked.
   * Enforces IMAGE_MAX_BYTES.
   */
  async readFileBytes(
    filePath: string,
  ): Promise<{ bytes: Buffer; mimeType: string; sizeBytes: number; mtimeMs: number }> {
    const { mimeType, sizeBytes, mtimeMs } = await this.getImageMeta(filePath);
    const bytes = await fs.readFile(filePath);

    return { bytes, mimeType, sizeBytes, mtimeMs };
  }

  /**
   * Write or overwrite file content.
   */
  async writeFileContent(filePath: string, content: string): Promise<string> {
    await fs.mkdir(path.dirname(filePath), { recursive: true });
    await Bun.write(filePath, content);
    return "File written successfully.";
  }

  /**
   * Delete a file from disk.
   */
  async deleteFile(filePath: string): Promise<boolean> {
    await fs.unlink(filePath);
    return true;
  }

  /**
   * Create a new directory.
   */
  async createDirectory(dirPath: string): Promise<boolean> {
    await fs.mkdir(dirPath, { recursive: true });
    return true;
  }

  /**
   * Delete a directory recursively.
   */
  async deleteDirectory(dirPath: string): Promise<boolean> {
    await fs.rm(dirPath, { recursive: true, force: true });
    return true;
  }
}

File: apps/server-go/internal/services/fs_service.go
// Filesystem operations. Port of apps/server/api/src/services/fs.service.ts
// and the preview gating rules in packages/types/src/fs.ts.
package services

import (
	"bytes"
	"fmt"

	"mime"
	"os"
	"path/filepath"
	"sort"
	"strconv"
	"strings"

	"github.com/Adelodunpeter25/console/apps/server-go/internal/fff"
	"github.com/Adelodunpeter25/console/apps/server-go/internal/types"
	"github.com/Adelodunpeter25/console/apps/server-go/internal/utils"
)

const (
	maxFilePreviewBytes = 512 * 1024
	imageMaxBytes       = 10 * 1024 * 1024
)

// manager is set once at startup; nil means fff is not wired.
var manager *fff.Manager

// SetFffManager wires the fff index manager into the fs service.
func SetFffManager(m *fff.Manager) { manager = m }

// PreviewBlocked carries the structured code/status the /file routes emit.
type PreviewBlocked struct {
	Code   string // LOCKFILE_BLOCKED | BINARY_FILE | FILE_TOO_LARGE
	Status int
	Detail map[string]any
	Msg    string
}

func (e *PreviewBlocked) Error() string { return e.Msg }

func BuildFileETag(sizeBytes int64, mtimeMs int64) string {
	return fmt.Sprintf("W/\"%s-%s\"",
		strconv.FormatInt(sizeBytes, 36),
		strconv.FormatInt(mtimeMs, 36))
}

var lockFileBasenames = map[string]bool{
	"package-lock.json": true, "npm-shrinkwrap.json": true,
	"pnpm-lock.yaml": true, "composer.lock": true,
}

var binaryExtensions = map[string]bool{
	".png": true, ".jpg": true, ".jpeg": true, ".gif": true, ".webp": true, ".bmp": true,
	".ico": true, ".icns": true, ".tiff": true,
	".mp4": true, ".mov": true, ".avi": true, ".mkv": true, ".webm": true,
	".mp3": true, ".wav": true, ".flac": true, ".ogg": true, ".m4a": true,
	".zip": true, ".tar": true, ".gz": true, ".tgz": true, ".bz2": true, ".xz": true,
	".7z": true, ".rar": true, ".jar": true, ".war": true,
	".woff": true, ".woff2": true, ".ttf": true, ".otf": true, ".eot": true,
	".pdf": true, ".exe": true, ".dll": true, ".so": true, ".dylib": true,
	".lock": true, ".lockb": true,
}

var imageExtensions = map[string]bool{
	".png": true, ".jpg": true, ".jpeg": true, ".gif": true, ".webp": true,
	".bmp": true, ".ico": true, ".svg": true,
}

type FsService struct{}

func NewFsService() *FsService { return &FsService{} }

// checkPreviewGate returns a *PreviewBlocked when the file cannot be shown.
func checkPreviewGate(path string, size int64) *PreviewBlocked {
	base := filepath.Base(path)
	ext := strings.ToLower(filepath.Ext(path))
	if lockFileBasenames[base] || ext == ".lock" || ext == ".lockb" {
		return &PreviewBlocked{Code: "LOCKFILE_BLOCKED", Status: 403,
			Msg: fmt.Sprintf("Machine-generated lockfile %q cannot be previewed.", base)}
	}
	if binaryExtensions[ext] {
		return &PreviewBlocked{Code: "BINARY_FILE", Status: 403,
			Msg: fmt.Sprintf("Binary file %q cannot be previewed as text.", base)}
	}
	if size > maxFilePreviewBytes {
		return &PreviewBlocked{Code: "FILE_TOO_LARGE", Status: 413,
			Msg:    fmt.Sprintf("%q is %s — previews are capped at %s.", base, formatBytes(size), formatBytes(maxFilePreviewBytes)),
			Detail: map[string]any{"sizeBytes": size, "maxBytes": maxFilePreviewBytes}}
	}
	return nil
}

func formatBytes(n int64) string {
	const unit = 1024
	if n < unit {
		return fmt.Sprintf("%d B", n)
	}
	div, exp := int64(unit), 0
	for m := n / unit; m >= unit; m /= unit {
		div *= unit
		exp++
	}
	return fmt.Sprintf("%.1f %cB", float64(n)/float64(div), "KMGTPE"[exp])
}

// BrowseDirectory lists one directory for the file picker UI.
func (s *FsService) BrowseDirectory(targetPath string, showHidden bool) (types.FsBrowseResult, error) {
	home, _ := os.UserHomeDir()
	resolved := targetPath
	if resolved == "" {
		resolved = home
	}
	resolved, err := filepath.Abs(resolved)
	if err != nil {
		return types.FsBrowseResult{}, err
	}
	parent := filepath.Dir(resolved)
	var parentPath *string
	if parent != resolved {
		parentPath = &parent
	}

	dirEntries, err := os.ReadDir(resolved)
	if err != nil {
		return types.FsBrowseResult{}, err
	}
	entries := make([]types.FsTreeEntry, 0)
	for _, de := range dirEntries {
		if !showHidden && utils.IsHiddenName(de.Name()) {
			continue
		}
		entry := types.FsTreeEntry{
			Name:  de.Name(),
			Path:  filepath.Join(resolved, de.Name()),
			IsDir: de.IsDir(),
		}
		if !de.IsDir() && !utils.IsPathIgnored(de.Name()) {
			if info, err := de.Info(); err == nil {
				size := info.Size()
				entry.Size = &size
			}
		}
		entries = append(entries, entry)
	}
	sortEntries(entries)
	return types.FsBrowseResult{CurrentPath: resolved, ParentPath: parentPath, Entries: entries}, nil
}

func sortEntries(entries []types.FsTreeEntry) {
	sort.Slice(entries, func(i, j int) bool {
		if entries[i].IsDir != entries[j].IsDir {
			return entries[i].IsDir
		}
		return numericLess(entries[i].Name, entries[j].Name)
	})
}

// numericLess approximates locale-aware numeric alphabetical ordering.
func numericLess(a, b string) bool {
	return strings.Compare(a, b) < 0
}

// EntriesOptions is the exported form used by route handlers.
type EntriesOptions struct {
	WithSizes  bool
	MaxEntries int
}

// ListAllEntries walks a tree DFS-style: directories first, each directory
// immediately followed by its subtree, capped at maxEntries.
func (s *FsService) ListAllEntries(targetPath string, maxDepth int, showHidden bool, opts EntriesOptions) ([]types.FsTreeEntry, error) {
	resolved, err := filepath.Abs(targetPath)
	if err != nil {
		return nil, err
	}
	if opts.MaxEntries <= 0 {
		opts.MaxEntries = 30000
	}
	counter := 0
	truncated := false
	var walk func(dir string, depth int) []types.FsTreeEntry
	walk = func(dir string, depth int) []types.FsTreeEntry {
		if depth > maxDepth || truncated {
			return nil
		}
		dirEntries, err := os.ReadDir(dir)
		if err != nil {
			return nil
		}
		visible := make([]os.DirEntry, 0, len(dirEntries))
		for _, de := range dirEntries {
			if !showHidden && utils.IsHiddenName(de.Name()) {
				continue
			}
			if !showHidden && utils.IsPathIgnored(de.Name()) {
				continue
			}
			visible = append(visible, de)
		}
		sort.Slice(visible, func(i, j int) bool {
			if visible[i].IsDir() != visible[j].IsDir() {
				return visible[i].IsDir()
			}
			return numericLess(visible[i].Name(), visible[j].Name())
		})

		local := make([]types.FsTreeEntry, 0, len(visible))
		for _, de := range visible {
			if truncated || counter >= opts.MaxEntries {
				truncated = true
				break
			}
			counter++
			entry := types.FsTreeEntry{
				Name:  de.Name(),
				Path:  filepath.Join(dir, de.Name()),
				IsDir: de.IsDir(),
			}
			if opts.WithSizes && !de.IsDir() {
				if info, err := de.Info(); err == nil {
					size := info.Size()
					entry.Size = &size
				}
			}
			local = append(local, entry)
		}
		// Interleave children DFS-style after their parent directory.
		for i, entry := range local {
			if !entry.IsDir {
				continue
			}
			children := walk(entry.Path, depth+1)
			local[i].Children = children
			if len(children) > 0 {
				out := make([]types.FsTreeEntry, 0, len(local)+len(children))
				out = append(out, local[:i+1]...)
				out = append(out, children...)
				out = append(out, local[i+1:]...)
				local = out
			}
		}
		return local
	}
	return walk(resolved, 1), nil
}

// GetDirectoryTree renders the list-dir tool's tree format.
func (s *FsService) GetDirectoryTree(targetPath string, maxDepth int, showHidden bool) (string, error) {
	entries, err := s.ListAllEntries(targetPath, maxDepth, showHidden, EntriesOptions{WithSizes: false, MaxEntries: 3000})
	if err != nil {
		return "", err
	}
	depthNote := fmt.Sprintf(" (recursive, max depth %d)", maxDepth)
	header := fmt.Sprintf("Directory: %s%s\n", targetPath, depthNote)
	if len(entries) == 0 {
		return header + "(empty directory)", nil
	}
	var buf bytes.Buffer
	renderTree(&buf, entries, "")
	return header + buf.String(), nil
}

func renderTree(buf *bytes.Buffer, entries []types.FsTreeEntry, prefix string) {
	for i, entry := range entries {
		isLast := i == len(entries)-1
		connector := "├── "
		childPrefix := "│   "
		if isLast {
			connector = "└── "
			childPrefix = "    "
		}
		if entry.IsDir {
			buf.WriteString(prefix + connector + entry.Name + "/\n")
			if len(entry.Children) == 0 {
				buf.WriteString(prefix + childPrefix + "(empty)\n")
			} else {
				renderTree(buf, entry.Children, prefix+childPrefix)
			}
		} else {
			sizeStr := ""
			if entry.Size != nil {
				sizeStr = " (" + formatBytes(*entry.Size) + ")"
			}
			buf.WriteString(prefix + connector + entry.Name + sizeStr + "\n")
		}
	}
}

type FileContentMeta struct {
	Content   string `json:"content"`
	SizeBytes int64  `json:"sizeBytes"`
	MtimeMs   int64  `json:"mtimeMs"`
}

// ReadFileContentWithMeta reads text content with preview gating.
func (s *FsService) ReadFileContentWithMeta(path string, startLine, endLine int) (FileContentMeta, error) {
	info, err := os.Stat(path)
	if err != nil {
		return FileContentMeta{}, err
	}
	if blocked := checkPreviewGate(path, info.Size()); blocked != nil {
		return FileContentMeta{}, blocked
	}
	raw, err := os.ReadFile(path)
	if err != nil {
		return FileContentMeta{}, err
	}
	content := sliceLines(string(raw), startLine, endLine)
	return FileContentMeta{
		Content:   content,
		SizeBytes: info.Size(),
		MtimeMs:   info.ModTime().UnixMilli(),
	}, nil
}

// sliceLines extracts [startLine, endLine] (1-based, inclusive), matching
// the TS implementation's semantics.
func sliceLines(text string, startLine, endLine int) string {
	if startLine == 0 && endLine == 0 {
		return text
	}
	lines := strings.Split(text, "\n")
	count := len(lines)
	start0 := startLine
	if start0 < 1 {
		start0 = 1
	}
	endExcl := endLine
	if endExcl <= 0 || endExcl > count {
		endExcl = count
	}
	if start0-1 >= count || endExcl <= start0-1 {
		return ""
	}
	return strings.Join(lines[start0-1:endExcl], "\n")
}

type ImageMeta struct {
	MimeType  string `json:"mimeType"`
	SizeBytes int64  `json:"sizeBytes"`
	MtimeMs   int64  `json:"mtimeMs"`
}

// GetImageMeta validates and describes an image file for raw preview.
func (s *FsService) GetImageMeta(path string) (ImageMeta, error) {
	ext := strings.ToLower(filepath.Ext(path))
	if !imageExtensions[ext] {
		return ImageMeta{}, &PreviewBlocked{Code: "BINARY_FILE", Status: 403,
			Msg: fmt.Sprintf("%q is not a previewable image.", filepath.Base(path))}
	}
	info, err := os.Stat(path)
	if err != nil {
		return ImageMeta{}, err
	}
	if info.Size() > imageMaxBytes {
		return ImageMeta{}, &PreviewBlocked{Code: "FILE_TOO_LARGE", Status: 413,
			Msg:    fmt.Sprintf("%q is %s — image previews are capped at %s.", filepath.Base(path), formatBytes(info.Size()), formatBytes(imageMaxBytes)),
			Detail: map[string]any{"sizeBytes": info.Size(), "maxBytes": imageMaxBytes}}
	}
	mimeType := mime.TypeByExtension(ext)
	if ext == ".svg" {
		mimeType = "image/svg+xml"
	}
	return ImageMeta{MimeType: mimeType, SizeBytes: info.Size(), MtimeMs: info.ModTime().UnixMilli()}, nil
}

// WriteFileContent creates or overwrites a file (mkdir -p the parent).
func (s *FsService) WriteFileContent(path, content string) (string, error) {
	if dir := filepath.Dir(path); dir != "" {
		if err := os.MkdirAll(dir, 0o755); err != nil {
			return "", err
		}
	}
	if err := os.WriteFile(path, []byte(content), 0o644); err != nil {
		return "", err
	}
	return fmt.Sprintf("Wrote %s (%d bytes)", path, len(content)), nil
}

func (s *FsService) DeleteFile(path string) (bool, error) {
	if err := os.Remove(path); err != nil {
		return false, err
	}
	return true, nil
}

func (s *FsService) CreateDirectory(path string) (bool, error) {
	if err := os.MkdirAll(path, 0o755); err != nil {
		return false, err
	}
	return true, nil
}

func (s *FsService) DeleteDirectory(path string) (bool, error) {
	if err := os.RemoveAll(path); err != nil {
		return false, err
	}
	return true, nil
}

// SearchFiles serves /api/fs/search. When the fff C library is available the
// fuzzy search runs through its index (much faster, frecency-ranked); until
// the index is warm — and always when fff is absent — it falls back to a
// substring walk, mirroring the TS fff-node semantics. Items use the TS
// FileSearchResult shape (relative/absolute paths plus score).
func (s *FsService) SearchFiles(root, query string, limit int, includeDirs bool) ([]types.FileSearchResult, error) {
	if manager != nil && manager.Enabled() {
		if items, ok := manager.SearchAsync(root, query, limit); ok {
			resolved, _ := filepath.Abs(root)
			out := make([]types.FileSearchResult, 0, len(items))
			for _, item := range items {
				full := filepath.Join(resolved, item.RelPath)
				out = append(out, types.FileSearchResult{
					RelativePath: item.RelPath, AbsolutePath: full, IsDir: item.IsDir,
				})
			}
			return out, nil
		}
	}
	resolved, err := filepath.Abs(root)
	if err != nil {
		return nil, err
	}
	if limit <= 0 {
		limit = 20
	}
	q := strings.ToLower(query)
	out := make([]types.FileSearchResult, 0)
	addItem := func(abs string, isDir bool) bool {
		rel, err := filepath.Rel(resolved, abs)
		if err != nil {
			rel = abs
		}
		out = append(out, types.FileSearchResult{RelativePath: rel, AbsolutePath: abs, IsDir: isDir})
		return len(out) >= limit
	}
	_ = filepath.WalkDir(resolved, func(path string, d os.DirEntry, err error) error {
		if err != nil {
			return nil
		}
		if d.IsDir() {
			if path != resolved && (utils.IsPathIgnored(d.Name()) || (!includeDirs && false)) {
				return filepath.SkipDir
			}
			if utils.IsHiddenName(d.Name()) && path != resolved {
				return filepath.SkipDir
			}
			if includeDirs && q != "" && strings.Contains(strings.ToLower(d.Name()), q) && path != resolved {
				if addItem(path, true) {
					return filepath.SkipAll
				}
			}
			return nil
		}
		if utils.IsPathIgnored(d.Name()) {
			return nil
		}
		if q != "" {
			// Separator-insensitive comparison so fallback still matches
			// camel/underscore queries like "ptymanager" -> pty_manager.go.
			lq, lp := strings.ToLower(q), strings.ToLower(path[len(resolved)+1:])
			lq = strings.Map(func(r rune) rune {
				if r == '_' || r == '-' || r == '/' || r == '.' {
					return -1
				}
				return r
			}, lq)
			lp = strings.Map(func(r rune) rune {
				if r == '_' || r == '-' || r == '/' || r == '.' {
					return -1
				}
				return r
			}, lp)
			if !strings.Contains(lp, lq) {
				return nil
			}
		}
		if addItem(path, false) {
			return filepath.SkipAll
		}
		return nil
	})
	return out, nil
}

// GrepOptions carries the /api/fs/grep query params, mirroring the global
// search panel's Aa/ab/.* toggles.
type GrepOptions struct {
	Mode         fff.GrepMode
	Case         fff.CaseMode
	WholeWord    bool
	ContextLines int
	MaxMatches   int
	Cursor       uint32
}

// ErrFffUnavailable is returned when the fff index isn't wired in for this
// server (e.g. missing native library), so the caller can render an empty
// state instead of a hard error.
var ErrFffUnavailable = fmt.Errorf("content search index is unavailable")

// Grep serves /api/fs/grep — content search across a workspace root for the
// global search panel (⌘⇧F). Requires the fff index; there is no filesystem
// walk fallback here since the panel is a fast/interactive UI, not the
// agent's best-effort tool.
func (s *FsService) Grep(root, query string, opts GrepOptions) (types.GrepResult, error) {
	if manager == nil || !manager.Enabled() {
		return types.GrepResult{}, ErrFffUnavailable
	}
	resolvedRoot, err := filepath.Abs(root)
	if err != nil {
		return types.GrepResult{}, err
	}
	lease, err := manager.GetOrCreate(resolvedRoot)
	if err != nil {
		return types.GrepResult{}, err
	}
	defer lease.Release()

	result, err := lease.GrepWithOptions(query, fff.GrepOptions{
		Mode:         opts.Mode,
		Case:         opts.Case,
		WholeWord:    opts.WholeWord,
		ContextLines: opts.ContextLines,
		MaxMatches:   opts.MaxMatches,
		Cursor:       opts.Cursor,
	})
	if err != nil {
		return types.GrepResult{}, err
	}

	matches := make([]types.GrepMatch, 0, len(result.Matches))
	for _, m := range result.Matches {
		ranges := make([]types.GrepMatchRange, 0, len(m.MatchRanges))
		for _, r := range m.MatchRanges {
			ranges = append(ranges, types.GrepMatchRange{Start: r.Start, End: r.End})
		}
		matches = append(matches, types.GrepMatch{
			RelPath:      m.RelPath,
			FileName:     m.FileName,
			LineNumber:   m.LineNumber,
			Column:       m.Column,
			EndColumn:    m.EndColumn,
			LineContent:  m.LineContent,
			MatchRanges:  ranges,
			IsBinary:     m.IsBinary,
			IsDefinition: m.IsDefinition,
		})
	}

	out := types.GrepResult{
		Matches:       matches,
		TotalMatched:  result.TotalMatched,
		FilesSearched: result.FilesSearched,
		TotalFiles:    result.TotalFiles,
		FilteredFiles: result.FilteredFiles,
		NextCursor:    result.NextCursor,
		HasMore:       result.HasMore,
	}
	if result.RegexError != nil {
		out.RegexError = result.RegexError.Message
	}
	return out, nil
}
